src/board.rs

```diff
@@ -39,83 +41,89 @@ impl Board {
+    pub fn is_in_bounds(position: Position) -> ChessResult<()> {
         if position.x > 7 || position.y > 7 {
             Err(ChessError::OutOfBounds)
         } else {
             Ok(())
         }
     }
 
-    fn is_valid_move(&self, from: Position, to: Position) -> Result<(), ChessError> {
-        Self::is_in_bounds(from)?;
-        Self::is_in_bounds(to)?;
-
-        let piece = &self.squares[from.x][from.y];
-
-        match *piece {
-            Piece::Pawn(..) => {}
-            Piece::Knight(..) => {}
-            Piece::Bishop(..) => {}
-            Piece::Rook(..) => {}
-            Piece::Queen(..) => {}
-            Piece::King(..) => {}
-            Piece::None => return Err(ChessError::InvalidMove),
-        };
-        Ok(())
+    fn is_piece_some(&self, position: Position) -> ChessResult<()> {
+        if self.get_piece(position)?.is_none() {
+            Err(ChessError::NoPieceAtPosition)
+        } else {
+            Ok(())
+        }
     }
 
-    pub fn move_piece(&mut self, from: Position, to: Position) -> Result<(), ChessError> {
-        // Check if 'from' and 'to' positions are in bounds
-        Self::is_in_bounds(from)?;
+    fn is_move_valid(&self, from: Position, to: Position) -> ChessResult<()> {
+        //we need ending position to be inbound
         Self::is_in_bounds(to)?;
 
-        if self._get_piece(from).is_none() {
-            return Err(ChessError::NoPieceAtPosition);
+        //moving piece must be in bound and not `None`
+        self.is_piece_some(from)?;
+
+        if let Some(piece) = self._get_piece(from) {
+            match piece {
+                Piece::Pawn(..) => {}
+                Piece::Knight(..) => {}
+                Piece::Bishop(..) => {}
+                Piece::Rook(..) => {}
+                Piece::Queen(..) => {}
+                Piece::King(..) => {}
+            };
+            Ok(())
+        } else {
+            Err(ChessError::NoPieceAtPosition)
         }
-        // Check if move is valid
-        self.is_valid_move(from, to)?;
+    }
+
+    pub fn move_piece(&mut self, from: Position, to: Position) -> ChessResult<()> {
+        self.is_piece_some(from)?;
+        self.is_move_valid(from, to)?;
 
-        // Perform the move
         self.squares[to.x][to.y] = self.squares[from.x][from.y].take();
 
         Ok(())
     }
 
-    pub fn get_piece(&self, position: Position) -> Result<Piece, ChessError> {
+    pub fn get_piece(&self, position: Position) -> ChessResult<Option<Piece>> {
         Self::is_in_bounds(position)?;
         Ok(self._get_piece(position))
     }
 
-    fn _get_piece(&self, position: Position) -> Piece {
+    fn _get_piece(&self, position: Position) -> Option<Piece> {
         self.squares[position.x][position.y]
     }
 
-    fn add_moves_in_direction(&mut self, start: Position, direction: Position) {
+    fn add_moves_in_direction(&mut self, start: Position, direction: Position) -> ChessResult<()> {
```

Comment:
We don't wan't to throw an error if we see an empty square. It's a normal part of this function.

Reply:
So does this still need to return a `ChessResult`?